telegram-deterministic-keys-wallet-test/README.md

```diff
@@ -0,0 +1,144 @@
+- The signed user operation is displayed in the UI.
+
+## Error Handling and Feedback
+
+- **Encryption/Decryption Errors**: If the encryption or decryption fails (e.g., invalid password), an error message is shown to the user.
+- **Key Generation Errors**: If the key generation fails, the UI will display an error message and prevent further actions.
+- **Signing Errors**: If signing the user operation fails (e.g., missing or incorrect mnemonic), an error message will appear.
+
+## Dependencies
+
+- `ethers`: A library used for working with Ethereum and its wallet system, including signing messages and transactions.
+- `bip32`: A library that implements BIP32 (Hierarchical Deterministic Wallets) for deriving keys.
+- `bip39`: A library that implements BIP39 for mnemonic generation and seed derivation.
+- `tiny-secp256k1`: A small library for elliptic curve cryptography used for signing messages and transactions with secp256k1.
+
+## Disadvantage of Telegram
+
+### **1. Storage Limitations on Telegram Mini Apps**
+
+Telegram Mini Apps are designed to be lightweight and integrated directly into the Telegram client. However, they have some limitations when it comes to storing data persistently:
+
+- **LocalStorage**: Data saved using `localStorage` in a Telegram Mini App is stored on the client's browser or app environment. However, this data is **not persistent** when the app is closed, and it gets wiped if the app is restarted or the user switches devices. This behavior is similar to how browser `localStorage` works, making it unsuitable for storing critical or long-term data like private keys or user preferences.
+
+- **Lack of secure storage**: Telegram Mini Apps do not offer built-in encryption or secure storage mechanisms like other platforms (e.g., iOS Keychain or Android Keystore). Sensitive information such as private keys should be stored securely, but currently, there are no built-in solutions for securely storing such information in Telegram Mini Apps.
+
+### **2. Alternative Solutions for Consistent Storage**
+
+Given the limitations of Telegram’s storage mechanisms, here are some potential alternatives to achieve consistency and persistent storage in Telegram Mini Apps:
+
+#### **a. Use Server-Side Storage (Encrypted Backup)**
+One of the most reliable methods for persistent storage is to store sensitive data like private keys or user preferences on your own secure server. Here's how you can achieve this:
+1. **Encryption**: Always encrypt sensitive data before sending it to the server. Use algorithms like AES to encrypt the data.
+2. **Server Storage**: Once encrypted, store the data on your backend. Ensure that your server is protected by secure protocols (e.g., HTTPS, token-based authentication).
+3. **Retrieve Data**: When the user reopens the Mini App, fetch the encrypted data from your server and decrypt it on the client-side.
+
+#### **b. Use Telegram's Bot API for Persistent Storage**
```

Comment:
Can you provide more info or link regarding this, I didn't find this kind of storage in their docs? It would be good to be covered also https://core.telegram.org/bots/webapps#cloudstorage

Reply:
Thanks for your review! 
I meant that the Bot API can work with external databases like MongoDB or Firebase for storing data. 

However, I think CloudStorage should be considered an alternative for managing data. I’ll update the documentation to reflect this and include a reference to the [CloudStorage docs](https://core.telegram.org/bots/webapps#cloudstorage).